### DISCUSSION TITLE
How to turn edges into lines by autotrace functionality?

### DISCUSSION THREAD

**Discussion by keto33** [ImageMagick/ImageMagick#2095] [Help]
I read an amazing outlook by @snibgo (L[ines, points and curves](http://www.im.snibgo.com/linptcv.htm)) on transferring points into bezier curves. I am struggling with the exact same problem. I could not directly implement the codes, as they are written for Windows, which I am unfamiliar with, as a Linux user. I tried to compile a C++ script given (snibFitCurve.cpp) as a starting point to reinvent the wheel, but apparently, I lack the required libraries (e.g., ArgType.h, though I downloaded the ones instructed).

As pointed out by @snibgo, autotrace -centerline option is the best choice, but the lines are discontinued at the pixel level. -line-threshold controls the smoothness of the bezier curves, but does not join disconnected lines.

I wonder if there is a possibility to get better results by using autotrace as an internal delegate of ImageMagick?

```
Version: ImageMagick 7.0.10-16 Q16 x86_64 2020-06-01 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(4.5) 
Delegates (built-in): autotrace bzlib djvu fftw fontconfig freetype jbig jng jp2 jpeg lcms lqr ltdl lzma openexr pangocairo png tiff webp wmf x xml zlib
```

**Comment by snibgo** [comment]
Oops, sorry, I've uploaded ArgType.h to http://snibgo.com/imforums/ArgType.h

You probably need just the BOOL stuff.

Since I wrote that stuff in 2016, other people may have tackled and solved similar problems.

**Reply by keto33** [reply]
I wish you were right, but it seems you were ahead of time. I searched a lot to find a similar library (e.g., here in github), but no luck even now in 2020.

The reason I gave up compiling the C++ script was my limited knowledge of C++ to re-write the missing libraries. Now, **FatalErr.h** is missing :)

**Comment by snibgo** [comment]
FatalErr.cpp calls other stuff, in a rat's nest. I suggest you write your own: FatalError() printf an error message and exit(1) the program. WarningError() does the same without the exit(1).

**Comment by mrx23dot** [comment]
What you are looking for is pixel art upscaling, it fills out pixels for smooth lines:
`ffmpeg -i input_1bit.png -filter_complex hqx=4 out.png`
then autotrace will have easier work.
